datafusion/expr-common/src/signature.rs

```diff
@@ -227,25 +226,11 @@ impl Display for TypeSignatureClass {
 
 #[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Hash)]
 pub enum ArrayFunctionSignature {
-    /// Specialized Signature for ArrayAppend and similar functions
-    /// The first argument should be List/LargeList/FixedSizedList, and the second argument should be non-list or list.
-    /// The second argument's list dimension should be one dimension less than the first argument's list dimension.
-    /// List dimension of the List/LargeList is equivalent to the number of List.
-    /// List dimension of the non-list is 0.
-    ArrayAndElement,
-    /// Specialized Signature for ArrayPrepend and similar functions
-    /// The first argument should be non-list or list, and the second argument should be List/LargeList.
-    /// The first argument's list dimension should be one dimension less than the second argument's list dimension.
-    ElementAndArray,
-    /// Specialized Signature for Array functions of the form (List/LargeList, Index+)
-    /// The first argument should be List/LargeList/FixedSizedList, and the next n arguments should be Int64.
-    ArrayAndIndexes(NonZeroUsize),
-    /// Specialized Signature for Array functions of the form (List/LargeList, Element, Optional Index)
-    ArrayAndElementAndOptionalIndex,
-    /// Specialized Signature for ArrayEmpty and similar functions
-    /// The function takes a single argument that must be a List/LargeList/FixedSizeList
-    /// or something that can be coerced to one of those types.
-    Array,
+    /// A function takes at least one List/LargeList/FixedSizeList argument.
+    Array {
+        /// A full list of the arguments accepted by this function.
+        arguments: Vec<ArrayFunctionArgument>,
+    },
```

Comment:
One drawback of this structure is that someone can technically create a signature with no array, only for it to get rejected at runtime in `get_valid_types()`. An alternative structure could be:

```Rust
/// A list of arguments that come before the array.
pre_array_aguments: Vec<ArrayFunctionArgument>,
/// A list of arguments that come after the array.
post_array_aguments: Vec<ArrayFunctionArgument>,
```

Then we would remove the `ArrayFunctionArgument::Array` variant. This would have the draw back of only allowing a single array argument in the signature, which might be fine.

Reply:
Another option is to force people to create this through a constructor that returns an error if there's no array argument.